src/OmniSharp.MiscellanousFiles/MiscellanousFiles.cs

```diff
@@ -0,0 +1,115 @@
+        private readonly OmniSharpWorkspace _workspace;
+        private readonly IFileSystemWatcher _fileSystemWatcher;
+        private readonly FileSystemHelper _fileSystemHelper;
+        private readonly ProjectSystem _projectSystem;
+        private readonly ILogger _logger;
+
+        [ImportingConstructor]
+        public MiscellanousFiles(OmniSharpWorkspace workspace, IFileSystemWatcher fileSystemWatcher, FileSystemHelper fileSystemHelper,
+            ILoggerFactory loggerFactory, [ImportMany] IEnumerable<Lazy<IProjectSystem, ProjectSystemMetadata>> projectSystems)
+        {
+            _workspace = workspace;
+            _fileSystemWatcher = fileSystemWatcher ?? throw new ArgumentNullException(nameof(fileSystemWatcher));
+            _fileSystemHelper = fileSystemHelper;
+            _logger = loggerFactory.CreateLogger<MiscellanousFiles>();
+            _projectSystem = (ProjectSystem)projectSystems.Where(ps => ps.Metadata.Name == nameof(ProjectSystem)).First().Value;
+        }
+
+        Task<object> IProjectSystem.GetProjectModelAsync(string filePath)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        Task<object> IProjectSystem.GetWorkspaceModelAsync(WorkspaceInformationRequest request)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        void IProjectSystem.Initalize(IConfiguration configuration)
+        {
+            var allFiles = _fileSystemHelper.GetFiles("**/*.cs");
+            foreach (var filePath in allFiles)
+                AddIfMiscellanousFile(filePath);
+
+            _fileSystemWatcher.Watch(miscFileExtension, OnMiscellanousFileChanged);
+        }
+
+        private async void AddIfMiscellanousFile(string filePath)
+        {
+            //wait for the project system to get initialised
+            await _projectSystem.HasCompletedUpdateRequest();
+            if (_workspace.GetDocument(filePath) == null)
+            {
+                string assemblyName = Guid.NewGuid().ToString("N");
+                var projectInfo = ProjectInfo.Create(
+                   filePath: filePath,
+                   id: ProjectId.CreateNewId(),
+                   version: VersionStamp.Create(),
+                   name: Path.GetFileName(filePath),
+                   metadataReferences: new MetadataReference[] { MetadataReference.CreateFromFile((typeof(object).Assembly).Location) },
+                   assemblyName: assemblyName,
+                   language: Language);
+
+                _workspace.AddProject(projectInfo);
+                _workspace.AddMiscellanousFileDocument(projectInfo.Id, filePath);
+                _projects[filePath] = projectInfo;
+                _logger.LogInformation($"Successfully added file '{filePath}' to workspace");
+            }
+        }
+
+        private void OnMiscellanousFileChanged(string filePath, FileChangeType changeType)
```

Comment:
How does this interact with the BufferManager logic where new files get placed in an existing project?

Reply:
The MSBuildProjectSystem interacts with the BufferManager and if a file is added to a project that it is already watching, then the BufferManager is called to add the new document. Only once MSBuild has completed this update process, does the Miscellanous FIle system add a new file, because of this await here: https://github.com/OmniSharp/omnisharp-roslyn/pull/1252/files/dd73d39631d118010ea073878b54755555e5f4ea#diff-b0a00a01200d50b36022179ac672964dR71. Thoughts ?

Reply:
I'm not sure I understood your answer to my question. And your link sends me to a giant diff rather than a specific location. The BufferManager has code which checks to see any files that it is notified about (through a `/changeBuffer` or `/updateBuffer` request) are already included in the OmniSharpWorkspace. If a file is not found, that code tries to find a logical project to include it within. I suspect that there may be cases with the MiscellaneousFileProjectSystem will have already added the file to the the "misc files" project and the BufferManager will never place new files in the correct project. This will be a regression.